plasmapy/utils/decorators/helpers.py

```diff
@@ -1,12 +1,105 @@
 """
-__all__ = ["preserve_signature"]
+__all__ = ["modify_docstring", "preserve_signature"]
 
 import functools
 import inspect
 
 
+def modify_docstring(func=None, prepend: str = None, append: str = None):
+    """
+    A decorator which programmatically prepends and/or appends the docstring
+    of the decorated method/function.  The unmodified/original docstring is
+    saved as the ``__original_doc__`` attribute.
+
+    Parameters
+    ----------
+    func: callable
+        The method/function to be decorated.
+
+    prepend: `str`
+        The string to be prepended to the ``func``'s docstring.
+
+    append: `str`
+        The string to be appended to the ``func``'s docstring.
+
+    Returns
+    -------
+    callable
+        Wrapped version of the function.
+
+    Examples
+    --------
+
+        >>> @modify_docstring(prepend='''Hello''', append='''World''')
+        ... def foo():
+        ...     '''Beautiful'''
+        ...     pass
+        >>> foo.__original_doc__
+        'Beautiful'
+        >>> foo.__doc__
+        'Hello\\n\\nBeautiful\\n\\nWorld'
+
+    """
+
+    def decorator(f):
+        sig = inspect.signature(f)
+
+        @preserve_signature
+        @functools.wraps(f)
+        def wrapper(*args, **kwargs):
+            # combine args and kwargs into dictionary
+            bound_args = sig.bind(*args, **kwargs)
+            bound_args.apply_defaults()
+
+            return f(*bound_args.args, **bound_args.kwargs)
+
+        if prepend is None and append is None:
+            raise TypeError(
+                "Decorator @modify_docstring() missing argument 'prepend' and/or"
+                " 'append', at least one argument is required."
+            )
+
+        # save the original docstring
+        setattr(wrapper, "__original_doc__", wrapper.__doc__)
+        doclines = inspect.cleandoc(wrapper.__doc__).splitlines()
+
+        # prepend docstring lines
+        if isinstance(prepend, str):
+            prependlines = inspect.cleandoc(prepend).splitlines()
+            prependlines.append("")
+        elif prepend is None:
+            prependlines = []
+        else:
+            raise TypeError(
+                f"Expected type str for argument 'prepend', got {type(prepend)}."
+            )
+
+        # append docstring lines
+        if isinstance(append, str):
+            appendlines = inspect.cleandoc(append).splitlines()
+            appendlines = [""] + appendlines
+        elif append is None:
+            appendlines = []
+        else:
+            raise TypeError(
+                f"Expected type str for argument 'append', got {type(append)}."
+            )
+
+        # define new docstring
+        wrapper.__doc__ = "\n".join(prependlines + doclines + appendlines)
+
+        return wrapper
+
+    if func is not None:
+        # `modify_docstring` called as a function
+        return decorator(func)
+    else:
+        # `modify_docstring` called as a decorator "sugar-syntax"
+        return decorator
```

Comment:
This comes up for a whole bunch of decorators that have optional arguments, so I'm wondering if we should create a function for it.  Doesn't have to be now, though.

Reply:
Maybe for the future.  I do always look up one of our old decorators to remind me how to do it.